[Sema] Support additional args in @dynamicMemberLookup subscripts

itaiferber opened the pull request:
Adds support for satisfying `@dynamicMemberLookup` requirements using subscripts which have arguments following `dynamicMember:` so long as they are either variadic or have default arguments. This allows transforming member references into `x[dynamicMember:...]` calls which can pass in arguments such as `#function`, `#fileID`, `#line`, etc.

* Pitch: https://forums.swift.org/t/dynamicmemberlookup-subscript-with-additional-defaulted-arguments/78471
* Proposal: https://github.com/swiftlang/swift-evolution/pull/2814

Main changes:

1. Allows `SubscriptDecl`s to satisfy `@dynamicMemberLookup` requirements as long as `dynamicMember:` remains the first explicitly-labeled argument, and all following arguments are either `isDefaultArgument()` or `isVariadic()`
2. Caches the result of checking for eligibility in `SubscriptDecl.Bits`; this was being checked multiple times per decl and it seems prudent to just store the info since we have the bits
3. Updates `ExprRewriter` to produce `ArgumentList`s for these subscripts by filling in `DefaultArgumentExpr`s as needed
4. Adds unit tests

The PR has been split into individual commits for (ideally) ease of review.

itaiferber commented on include/swift/AST/Decl.h:
```diff
@@ -7381,6 +7400,36 @@ enum class ObjCSubscriptKind {
+///      a default value
+///
+/// Subscripts which don't meet these requirements strictly are not eligible.
+enum class DynamicMemberLookupSubscriptEligibility : uint8_t {
```
`DynamicMemberLookupSubscriptEligibility` is... a mouthful — but it's the least-inaccurate name I could come up with. Happy to improve this with suggestions!

itaiferber commented on include/swift/AST/Decl.h:
```diff
@@ -7496,6 +7573,16 @@ class SubscriptDecl : public GenericContext, public AbstractStorageDecl {
 
+  /// Determines, caches, and returns whether the decl can be used to satisfy an
+  /// `@dynamicMemberLookup` requirement (and if so, how).
+  DynamicMemberLookupSubscriptEligibility
+  getDynamicMemberLookupSubscriptEligibility();
```
Instead of exposing this enum, I suppose it is also possible to restore `isValidDynamicMemberLookupSubscript()` et. al., but I do think it's nicer to have a type-safe interface that's harder to typo.

itaiferber commented on lib/Sema/CSApply.cpp:
```diff
@@ -3594,13 +3631,68 @@ namespace {
+                                                      paramTy);
+          }
+        } else {
+          // If bindings were substituted, we need to find the "original" (or
+          // contextless) parameter index for the default argument.
+          if (shouldSubstituteParameterBindings(subscript)) {
+            auto *paramList = SD->getParameterList();
+            assert(paramList);
+            paramIdx = paramList->getOrigParamIndex(
+                subscript.getSubstitutions(), paramIdx);
+          }
+
+          auto owner = getDefaultArgOwner(subscript, paramIdx);
+          auto paramTy = params[paramIdx].getParameterType();
+          argExpr = new (ctx)
+              DefaultArgumentExpr(owner, paramIdx, componentLoc, paramTy, dc);
```
This is pretty much what happens in `coerceCallArguments`, but trying to DRY this out didn't end up feeling worth it.

itaiferber commented:
@swift-ci please smoke test

itaiferber committed:
Cache SubscriptDecl @dynamicMemberLookup eligibility

`SubscriptDecl`s may get checked multiple times for eligibility in
fulfilling `@dynamicMemberLookup` requirements; since the checks are
non-trivial and the result doesn't change, this eligibility can be
cached in the decl's `Bits`.

itaiferber committed:
Use SubscriptDecl interface for @dynamicMemberLookup checks

`SubscriptDecl` exposes eligibility for `@dynamicMemberLookup`
requirements directly, so the `TypeChecker` interface for these members
can be replaced.

itaiferber committed:
Support additional args in @dynamicMemberLookup subscripts

Adds support for `SubscriptDecl`s to fulfill `@dynamicMemberLookup`
requirements if they have additional arguments after `dynamicMember:` so
long as those arguments have default values, or are variadic.

This allows exposing values like `#function`, `#fileID`, `#line`, etc.
to dynamic member lookup.

itaiferber committed:
Add unit tests for multi-arg @dynamicMemberLookup subscripts

itaiferber commented:
@swift-ci Please smoke test

xwu commented:
@swift-ci Please build toolchain

itaiferber committed:
Add isolation info to ExtInfo for SubscriptDecls

Subscripts with `isolated` arguments currently do not pass a compiler
assertion about the consistency of the function type w.r.t. isolation.
This is because the `ExtInfo` for `SubscriptDecl`s doesn't currently
have isolation information applied (though this works in practice);
annotating the `ExtInfo` with isolation allows this assertion to pass.

slavapestov commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2060,215 +2063,182 @@ visitDynamicCallableAttr(DynamicCallableAttr *attr) {
-  if (ignoreLabel) {
-    return true;
+DynamicMemberLookupSubscriptEligibility
+SubscriptDecl::evaluateDynamicMemberLookupEligibility(
```
Did you decide against making this into a request for some reason?

Also since it caches the value, what does it mean if this is called with different AccessScopes?

hamishknight replied:
Yeah I think this should do something similar to `IsSingleValueStmtRequest`/`IsSingleValueStmtResult`

itaiferber replied:
Will do! Thanks for the pointer to `IsSingleValueStmtRequest` — I'll take a look at the pattern there.

slavapestov commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -2785,6 +2786,28 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
+                if (SD->evaluateDynamicMemberLookupEligibility(
+                        &accessScope, &SD->getDiags()) ==
+                    DynamicMemberLookupSubscriptEligibility::None) {
+                  SD->setInvalid();
```
Let's not add new calls to `setInvalid()`, it's kind of a bogus operation

xedin replied:
I think this is too late for this sort of check, it should be part of override checking.

slavapestov commented on test/attr/attr_dynamic_member_lookup_default_args.swift:
```diff
@@ -0,0 +1,157 @@
+  let dsohandle: UnsafeRawPointer
+
+  init(
+    fileID: String = #fileID,
```
Do we need to test non-magic default arguments too? Probably silly and pointless to do that, but it should still not crash the compiler ;)

slavapestov reviewed (approved):
I didn't review this too closely, but given the large number of new tests I'm fine with this going in once you address the follow-up remarks

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -10778,20 +10778,23 @@ performMemberLookup(ConstraintKind constraintKind, DeclNameRef memberName,
+              args->getLabel(0) == getASTContext().Id_dynamicMember;
+          if (!isDirectCallToSubscript) {
+            return OverloadChoice::getDynamicMemberLookup(
+                baseTy, SD, ctx.getIdentifier("subscript"),
```
This is fine, it just records a name of a dynamic member it found for book-keeping purposes because I think for string one it could be not only subscript.

xedin reviewed (changes requested):
I think before we can land this we need to put new selection logic behind an experimental flag and it would be great to address comments in CSApply and `evaluateDynamicMemberLookupEligibility` as well.

hamishknight commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1728,9 +1728,14 @@ ERROR(invalid_dynamic_member_lookup_type,none,
+     "'dynamicMember' argument must appear first in the argument list", ())
+NOTE(invalid_dynamic_member_subscript_missing_default, none,
+     "add a default value to this argument to satisfy the "
+     "'@dynamicMemberLookup' requirement", ())
```
Nit: these are parameters not arguments

```suggestion
NOTE(invalid_dynamic_member_subscript_out_of_order, none,
     "'dynamicMember' argument must appear first in the parameter list", ())
NOTE(invalid_dynamic_member_subscript_missing_default, none,
     "add a default value to this parameter to satisfy the "
     "'@dynamicMemberLookup' requirement", ())
```

hamishknight commented on lib/Sema/CSApply.cpp:
```diff
@@ -3629,13 +3632,63 @@ namespace {
       // Build and type check the string literal index value to the specific
       // string type expected by the subscript.
       auto *nameExpr = new (ctx) StringLiteralExpr(name, loc, /*implicit*/true);
+      nameExpr->setType(literalTy);
```
This shouldn't be necessary, during CSApply the ConstraintSystem is the source of truth for the types of the AST nodes, we then go through and set the types on the nodes when we're done (I know, it's weird, but it would be good to stick to the convention I think).

itaiferber commented on lib/Sema/CSApply.cpp:
```diff
@@ -3629,13 +3632,63 @@ namespace {
+    /// arguments.
+    ArgumentList *buildArgumentListForDynamicMemberLookupSubscript(
+        const SelectedOverload &overload, SourceLoc componentLoc,
+        ConstraintLocator *locator) {
```
Unfortunately, I spent an inordinate amount of time trying to get this behavior threaded through `coerceCallArguments` directly without success. It's been a few months so I'll need to page the various failure modes back into memory (and I'll report them back here), but from what I remember at the moment:

1. `coerceCallArguments` asserts that the constraint locator being used ends with `ApplyArgument`, but using that locator fails to find the right `memberLoc` as needed here; threading through multiple locators was fine, but awkward
2. `coerceCallArguments` implicitly expects `funcType` to be the overload's `adjustedOpenedFullType`, but we need to use the non-full type to preserve parameter packs so we can create default exprs for them
3. Building the default argument exprs inside of `coerceCallArguments` requires needing to substitute parameter bindings, but IIRC either the list of bindings was always empty, or I kept hitting assertions in trying to get them to compute (sorry, don't remember clearly at this point)

In the end, the only way I managed to get this to work (cleanly) was to extract this exact behavior out, use it to create a base argument list for subscripts, and then pass that base list through `coerceCallArguments` itself when building up the actual exprs.

I'll try to get you more concrete details, but this is an area I'd be very grateful for some direct help with if possible!

xedin replied:
It sounds like we need to adjust how callee locators are computed in `getCalleeLocator` for dynamic member subscripts and change how they are synthesized for argument expressions as well, this is a solver-side change rather than CSApply one.

itaiferber commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2057,215 +2063,182 @@ visitDynamicCallableAttr(DynamicCallableAttr *attr) {
+           index->getArgumentName() != ctx.Id_dynamicMember) &&
+          (!index->isDefaultArgument() && !index->isVariadic() &&
+           !isa<PackExpansionType>(
+               index->getInterfaceType()->getCanonicalType()))) {
```
Good call!

itaiferber committed:
Convert SubscriptDecl @dynamicMemberLookup checking to request

Adds a new DynamicMemberLookupSubscriptRequest type for evaluating and
caching the validity of a `SubscriptDecl`'s usage to fulfill a
`@dynamicMemberLookup` requirement for a specific usage.

itaiferber committed:
Improve @dynamicMemberLookup diagnostics

Instead of producing a single error (with notes) for most
`@dynamicMemberLookup` failures, introduce individual errors with
fix-its.

itaiferber committed:
Migrate @dynamicMemberLookup subscript override checking to OverrideMatcher

This validation was being performed too late in override checking, and
is appropriate to perform along with other checks in
`checkSingleOverride`.

itaiferber committed:
Add non-magic default arg to tests

Expand `Magic` unit test to include non-magic default arg to confirm it
also works as expected at runtime.

itaiferber commented:
@xedin, @slavapestov, @hamishknight Thanks again for taking a look at my initial attempt — appreciate you all! When you have some time, I think this is ready for another look to see if this is a step in the right direction. There's no rush at all, and I know Apple folks are largely off this week. If this is looking better, then I'll polish up the commits/history a bit.

hamishknight commented on include/swift/AST/Decl.h:
```diff
@@ -7544,6 +7546,7 @@ enum class ObjCSubscriptKind {
   Keyed
 };
 
+
```
Could you remove these unrelated whitespace-only diffs?

hamishknight commented on include/swift/AST/Decl.h:
```diff
@@ -7598,6 +7614,11 @@ class SubscriptDecl : public GenericContext, public AbstractStorageDecl {
     setIndices(Indices);
   }
 
+  /// Returns the given as a `BoundGenericType` if it is a
```
```suggestion
  /// Returns the given type as a `BoundGenericType` if it is a
```

hamishknight commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -5612,6 +5594,189 @@ class DesugarForEachStmtRequest
+           // `OptionSet` intentionally does not offer `==` in favor of
+           // requiring `containsOnly`.
+           std::equal(lhs.ParamFlags.begin(), lhs.ParamFlags.end(),
+                      rhs.ParamFlags.begin(), rhs.ParamFlags.end(),
```
You can use `llvm::equal` which just takes the containers directly

hamishknight commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3385,3 +3385,204 @@ SourceLoc PrettyPrintDeclRequest::evaluate(Evaluator &eval, const Decl *decl) co
+        outputError = true;
+      }
+
+      if (flags & InvalidParameterFlag::DynamicMemberInvalidArgumentLabel) {
```
Should this branch set `outputError` too?

hamishknight commented on lib/Sema/CSApply.cpp:
```diff
@@ -3636,13 +3638,63 @@ namespace {
+                                                  componentLoc, paramTy, dc);
+        }
+
+        if (!argExpr) {
```
Hamish: variadics?

hamishknight commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3385,3 +3385,204 @@ SourceLoc PrettyPrintDeclRequest::evaluate(Evaluator &eval, const Decl *decl) co
+  auto seenDynamicMemberParamLabel = false;
+  std::optional<DynamicMemberKind> kind = std::nullopt;
+  SmallVector<InvalidParameterFlags> paramFlags;
+  for (auto idx : range(0, indices->size())) {
```
```suggestion
  for (auto idx : indices(indices)) {
```

hamishknight commented on lib/Sema/CSApply.cpp:
```diff
@@ -3636,13 +3638,63 @@ namespace {
+          }
+        } else {
+          argExpr = new (ctx) DefaultArgumentExpr(subscript, paramIdx,
+                                                  componentLoc, paramTy, dc);
```
This doesn't seem like it will handle variadics correctly, did you get a chance to see if we can use `coerceCallArguments` to fill in the right arguments here?

hamishknight commented on include/swift/AST/Decl.h:
```diff
@@ -7659,6 +7680,35 @@ class SubscriptDecl : public GenericContext, public AbstractStorageDecl {
+  /// access control is being used consistently and that `decl` is appropriately
+  /// accessible, returning `false` if inaccessible.
+  bool isValidDynamicMemberLookupSubscript(
+      std::optional<const DeclContext *> useDC) const;
```
That being said, I'm not convinced we should be checking for accessibility here (and in the below queries), this will hamper the effectiveness of the request caching. I think only the logic in TypeCheckAttr needs to care about this anyway, I would keep the check there

hamishknight replied:
> Today, we don't really do any accessibility checking for these subscripts inside of the constraint solver (or anywhere else)

That doesn't match my understanding, we go through `ConstraintSystem::performMemberLookup` for this which initially tries with `includeInaccessibleMembers` set to `false` and only flips it to `true` if we're attempting fixes, and then marks such candidates as `MemberLookupResult::UR_Inaccessible`. Do you have an example of a case where we allow you to reference e.g a `private` dynamic member subscript?

Generally I think `isValidDynamicMemberLookupSubscript` should be determining "is this a valid dynamic member subscript from the nominal decl's point of view", it should be up to use sites to check for accessibility, same as it would be for a regular member reference

hamishknight reviewed (commented):
Sorry for taking so long to get back to this! It's definitely looking more on the right track

itaiferber commented on lib/IDE/SourceEntityWalker.cpp:
```diff
@@ -10,13 +10,14 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "swift/IDE/SourceEntityWalker.h"
```
Absolutely! 🙈

hamishknight commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -5612,6 +5594,189 @@ class DesugarForEachStmtRequest
+    return !DynamicMemberIdx && Kind &&
+           ParamFlags[0].toRaw() ==
+               llvm::to_underlying(
+                   InvalidParameterFlag::DynamicMemberMissingArgumentLabel) &&
```
Oh is the OptionSet not implicitly constructed if you just pass the flag to `containsOnly`? If not, this is fine

itaiferber commented on test/attr/attr_dynamic_member_lookup.swift:
```diff
@@ -414,6 +644,31 @@ func testOverrideSubscript(a: BaseClass, b: DerivedClassWithSetter) {
+// Overriding with a different default value is valid.
+class DerivedClassWithMissingDefault : BaseClass {
+  // expected-error @+1 {{'@dynamicMemberLookup' requires 'magic' to have a default value}} {{71-71= = <#Default#>}}
+  override subscript(dynamicMember member: String, magic: StaticString) -> String {
```
Oh, I misunderstood! I thought you were referring to the comment for the diagnostic itself — the comment describing the test class itself is wrong. Yeah, will fix!

itaiferber committed:
fixup! Support additional args in @dynamicMemberLookup subscripts

itaiferber committed:
fixup! Convert SubscriptDecl @dynamicMemberLookup checking to request

itaiferber committed:
fixup! Convert SubscriptDecl @dynamicMemberLookup checking to request

itaiferber committed:
fixup! Cache SubscriptDecl @dynamicMemberLookup eligibility

itaiferber committed:
fixup! Convert SubscriptDecl @dynamicMemberLookup checking to request

itaiferber committed:
fixup! Cache SubscriptDecl @dynamicMemberLookup eligibility

itaiferber committed:
fixup! Convert SubscriptDecl @dynamicMemberLookup checking to request

itaiferber committed:
fixup! Convert SubscriptDecl @dynamicMemberLookup checking to request

itaiferber committed:
fixup! Convert SubscriptDecl @dynamicMemberLookup checking to request

itaiferber committed:
fixup! Convert SubscriptDecl @dynamicMemberLookup checking to request

itaiferber committed:
fixup! Convert SubscriptDecl @dynamicMemberLookup checking to request

itaiferber committed:
fixup! Use SubscriptDecl interface for @dynamicMemberLookup checks

itaiferber committed:
fixup! Convert SubscriptDecl @dynamicMemberLookup checking to request

itaiferber committed:
fixup! Convert SubscriptDecl @dynamicMemberLookup checking to request

itaiferber committed:
fixup! Convert SubscriptDecl @dynamicMemberLookup checking to request

itaiferber committed:
fixup! fixup! Support additional args in @dynamicMemberLookup subscripts

itaiferber committed:
fixup! Convert SubscriptDecl @dynamicMemberLookup checking to request

itaiferber committed:
fixup! Support additional args in @dynamicMemberLookup subscripts

itaiferber committed:
fixup! Improve @dynamicMemberLookup diagnostics

itaiferber commented:
Closing now that #88480 has been merged. Many thanks again to @slavapestov for taking on the work!
